Core/TypeParser.cpp

```diff
@@ -9,100 +9,148 @@
-    { 'i', "int" },
-    { 'l', "long" },
-    { 'C', "unsigned char" },
-    { 'S', "unsigned short" },
-    { 'I', "unsigned int" },
-    { 'L', "unsigned long" },
-    { 'f', "float" },
-    { 'A', "uint8_t" },
-    { 'b', "BOOL" },
-    { 'B', "BOOL" },
-
-    { 'q', "NSInteger" },
-    { 'Q', "NSUInteger" },
-    { 'd', "CGFloat" },
-    { '*', "char *" },
-
-    { '@', "id" },
-    { ':', "SEL" },
-    { '#', "objc_class_t" },
-
-    { '?', "void*" },
-    { 'T', "void*" },
-};
-
-std::vector<std::string> TypeParser::parseEncodedType(const std::string& encodedType)
+std::vector<QualifiedNameOrType> TypeParser::parseEncodedType(const std::string& encodedType)
 {
-    std::vector<std::string> result;
+    std::vector<QualifiedNameOrType> result;
     int pointerDepth = 0;
 
-    for (size_t i = 0; i < encodedType.size(); ++i) {
-        char c = encodedType[i];
+    bool readingNamedType = false;
+    std::string namedType;
+    size_t readingStructDepth = 0;
+    std::string structType;
 
-        // K: For example, ^@ is a single type, "id*".
-        if (c == '^') {
-            pointerDepth++;
+    for (char c : encodedType) {
+
+        if (readingNamedType && c != '"') {
+            namedType.push_back(c);
+            continue;
+        } else if (readingStructDepth > 0 && c != '{' && c != '}') {
+            structType.push_back(c);
             continue;
         }
 
-        // Argument frame size and offset specifiers aren't relevant here; they
-        // should just be skipped.
         if (std::isdigit(c))
             continue;
 
-        if (auto it = TypeEncodingMap.find(c); it != TypeEncodingMap.end()) {
-            std::string encoding = it->second;
-            for (int j = pointerDepth; j > 0; j--)
-                encoding += "*";
-            pointerDepth = 0;
-            result.emplace_back(encoding);
+        QualifiedNameOrType nameOrType;
+        std::string qualifiedName;
+
+        switch (c) {
+        case '^':
+            pointerDepth++;
             continue;
-        }
 
-        // (Partially) handle quoted type names.
-        if (c == '"') {
-            while (encodedType[i] != '"')
-                i++;
+        case '"':
+            if (!readingNamedType) {
+                readingNamedType = true;
+                continue;
+            } else {
+                readingNamedType = false;
+                nameOrType.name = QualifiedName(namedType);
+                break;
+            }
+        case '{':
+            readingStructDepth++;
+            continue;
+        case '}':
+            readingStructDepth--;
 
-            // TODO: Emit real type names.
-            pointerDepth = 0;
-            result.emplace_back("void*");
+            if (readingStructDepth == 0) {
+                // TODO: Emit real struct types
+                nameOrType.type = Type::PointerType(8, Type::VoidType());
+                break;
+            }
+            else
+                continue;
```

Comment:
The `else` here can be removed, i.e. this case can just have `continue` at the bottom.